dm/syncer/validator_checkpoint.go

```diff
@@ -0,0 +1,436 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+	"github.com/pingcap/tidb-tools/pkg/filter"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/dm/pb"
+	"github.com/pingcap/tiflow/dm/dm/unit"
+	"github.com/pingcap/tiflow/dm/pkg/binlog"
+	"github.com/pingcap/tiflow/dm/pkg/conn"
+	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/cputil"
+	"github.com/pingcap/tiflow/dm/pkg/gtid"
+	"github.com/pingcap/tiflow/dm/syncer/dbconn"
+)
+
+type validateErrorState int
+
+const (
+	// todo: maybe move to some common place, dmctl may use it
+	newValidateErrorRow validateErrorState = iota
+	//nolint
+	ignoredValidateErrorRow
+	//nolint
+	resolvedValidateErrorRow
+)
+
+type validatorPersistHelper struct {
+	tctx              *tcontext.Context
+	cfg               *config.SubTaskConfig
+	db                *conn.BaseDB
+	dbConn            *dbconn.DBConn
+	validator         *DataValidator
+	schemaInitialized atomic.Bool
+
+	checkpointTableName    string
+	pendingChangeTableName string
+	errorChangeTableName   string
+	tableStatusTableName   string
+	revision               int64
+}
+
+func newValidatorCheckpointHelper(validator *DataValidator) *validatorPersistHelper {
+	cfg := validator.cfg
+	c := &validatorPersistHelper{
+		cfg:       cfg,
+		validator: validator,
+
+		checkpointTableName:    dbutil.TableName(cfg.MetaSchema, cputil.ValidatorCheckpoint(cfg.Name)),
+		pendingChangeTableName: dbutil.TableName(cfg.MetaSchema, cputil.ValidatorPendingChange(cfg.Name)),
+		errorChangeTableName:   dbutil.TableName(cfg.MetaSchema, cputil.ValidatorErrorChange(cfg.Name)),
+		tableStatusTableName:   dbutil.TableName(cfg.MetaSchema, cputil.ValidatorTableStatus(cfg.Name)),
+	}
+
+	return c
+}
+
+func (c *validatorPersistHelper) init(tctx *tcontext.Context) error {
+	c.tctx = tctx
+
+	newCtx, cancelFunc := c.tctx.WithTimeout(unit.DefaultInitTimeout)
+	defer cancelFunc()
+
+	dbCfg := c.cfg.To
+	dbCfg.RawDBCfg = config.DefaultRawDBConfig().SetReadTimeout(maxDMLConnectionTimeout)
+	db, conns, err := dbconn.CreateConns(newCtx, c.cfg, &dbCfg, 1)
+	if err != nil {
+		return err
+	}
+
+	c.db = db
+	c.dbConn = conns[0]
+	defer func() {
+		if err == nil {
+			return
+		}
+		dbconn.CloseBaseDB(newCtx, c.db)
+	}()
+
+	if !c.schemaInitialized.Load() {
+		if err = c.createSchema(newCtx); err != nil {
+			return err
+		}
+
+		err = c.createTable(newCtx)
+
+		c.schemaInitialized.Store(true)
+	}
+	return err
+}
+
+func (c *validatorPersistHelper) createSchema(tctx *tcontext.Context) error {
```

Comment:
this schema and table should also removed in `remove-meta` ?

Reply:
add a todo